### ISSUE TITLE
Support for Tensorflow 2

### ISSUE THREAD

**Issue by faysou** [Bihaqo/t3f#190]
Is there any plan to migrate the library to support Tensorflow 2 ?
What would need to be changed in the library for this ?

**Comment by Bihaqo** [comment]
What level of support do you need?
Because you can always
   import tf.comp.v1 as tf
and be fine with the old code, but without new features.

**Comment by faysou** [comment]
Ok thank you, then this should be enough for me.
Great library by the way ! I've used it for tensor completion.
Tensor trains are quite magical, I knew about HOSVD, but this is big step forward.

**Comment by Bihaqo** [comment]
Thanks, glad someone uses it! :)

Note that you'll probably have to change the tf imports in the whole library (I know, annoying). I guess the simplest thing to do would be to fork the library and use "replace in all files" in some editor.

**Comment by Bihaqo** [comment]
I just realized that I probably can always import compat.v1 even in older version, so I'll try making it the default: #191

**Comment by faysou** [comment]
Yes that would be nice, so your library works out of the box with the newer version of tf.

**Comment by Bihaqo** [comment]
I think #191 should work, just don't forget to tf.disable_v2_behavior() (like in the tutorial: https://colab.research.google.com/github/Bihaqo/t3f/blob/tf2_dummy_support/docs/quick_start.ipynb)

**Comment by Bihaqo** [comment]
Will probably merge it tomorrow

**Comment by faysou** [comment]
I'm trying to use Session instead of eager in order to use gradients in an algorithm where the rank of a tensor is increased gradually. 

It seems hard in tensorflow 1 to avoid recomputations.

Does tensorflow 2 avoid this issue as the recommended mode in it is eager evaluation ? 
Maybe this library could become easier to use in tensorflow 2, as using tf.Session really complexify the code.

**Comment by Bihaqo** [comment]
Can you show a code example please? If feels like you shouldn't need
growing the ranks, but maybe I'm missing something :)

You can run t3f in eager mode, just do tf.enable_eager_execution() at the
top of your file. But not everything is supported, e.g. t3f.gradients
doesn't work in eager rmode ight now. It's not too hard to support it, but
I'm not sure when I'll have time to do that.

**Comment by Bihaqo** [comment]
I created a separate issue about it to not to forget: #193

**Comment by faysou** [comment]
Hi Alexander, sorry I didn't see your message.
You can have a look at the paper below, p14.

http://sma.epfl.ch/~anchpcommon/publications/ttcompletion.pdf

**Comment by Bihaqo** [comment]
Fait enough, I agree it would be easier to do in tf 2.

Note though that even in TF 2 it would be super annoying because you have to use tf.function to compile pieces of code to make it run faster, and tf.function doesn't support anything but tf tensors (i.e. it doesn't support t3f objects).

So you would write something like
```
def step(tens):
  # ... process
  # Do something that increases ranks
  tens = tens + 1
  return tens
for i in range(num_iter):
  tens = step(tens)
```

But then, to make it run reasonably fast you'll have to do use tf.function on step, but to do that you'll have to make it take a list of tf.Tensors as input and output a list of tf.Tensors, i.e. something like
```
@tf.function
def step(tens_cores):
  # ... process
  # Do something that increases ranks
  tens = t3f.TensorTrain(tens_cores)
  tens = tens + 1
  return tens.tt_cores
for i in range(num_iter):
  tens_cores = step(tens_cores)
```

And at this point you can do something like this with TF 1 as well, i.e.
```
def step(tens_cores):
  # ... process
  # Do something that increases ranks
  tens = t3f.TensorTrain(tens_cores)
  tens = tens + 1
  return tens.tt_cores
next_iter_cores = step(tens)
with tf.Session() as sess:
  for i in range(num_iter):
    tens_cores = sess.run(next_iter_cores, feed_dict={tens.tt_cores[i]: tens_cores[i] for i in 
 range(len(tens_cores))})
```

Anyway, it's a bit ugly in both TF 1 and TF 2, but TF 2 indeed would be nicer :)

I'll take a look at implementing t3f.gradients in eager mode.

**Comment by faysou** [comment]
Great, thank you for your reply.

**Comment by Bihaqo** [comment]
After thinking some more about it I'm less sure you can easely do that in
tf 1..

**Comment by faysou** [comment]
I'll wait for your v2 then :)

Your library could save a lot of energy in training, it should be used by more people for neural networks.

**Comment by Bihaqo** [comment]
Ok, done :)

This is not submitted yet though, so please checkout this branch: #193

**Comment by faysou** [comment]
Waw, thank you. So to be clear your new commit is about the auto-diff with eager mode, not full support for tensorflow 2.

I'll try your new commit it will normally speed up my code.

**Comment by Bihaqo** [comment]
Better support is added in #201
